### DISCUSSION TITLE
Is semihosting supported when running on an ST-Link/v2?

### DISCUSSION THREAD

**Discussion by Elias-Graf** [blackmagic-debug/blackmagic#1368] [Q&A] [answered]
I'm trying to follow the rust book for embedded programming (https://docs.rust-embedded.org/book/start/hardware.html). At some point it tells me to enable semihosting, which restuls in:

```
(gdb) monitor arm semihosting enable
Target does not support this command.
```

I'm new to this, so I have no clue if it's just not supported, if I compiled the blackmagic firmware wrong, if I compiled my program incorrectly, or screwed up something different entirely. I did see the https://black-magic.org/usage/semihosting.html section on the website, and I assume those flags need to be passed to the linker of my rust program. But I haven't found any way to do that / the rust book I'm reading does not mention this, nor did I find it in the documentation of https://crates.io/crates/cortex-m-semihosting (which implements the semihosting).

Help is greatly appreciated :blush:.

**Comment by dragonmux** [comment, marked as answer]
`monitor arm semihosting enable` is only for OpenOCD. BMP implements a different set of monitor commands (`monitor help` to list), and that we remember, semihosting should always be enabled once attached to an ARM core. Semihosting output goes to the secondary serial port for BMP (/dev/ttyBmpTarg if on Linux) when you run `monitor redirect_stdout enable`.

The ST-Link v2 platform should support semihosting, so the redirect command should be visible in the `monitor help` output.

You are correct that you do have to compile with an appropriate crate that invokes the semihosting halo using the break instruction that encodes to `0xbeab`, but that should be all that is required.

**Comment by Elias-Graf** [comment]
@dragonmux Ahh, thank you so much! Yes, I am on Linux and I see the `redirect_stdout` option when executing `monitor help`.

Just a little suggestion from a novice, wouldn't it be helpful if those commands were documented under https://black-magic.org/usage/semihosting.html? Because I thought it might work differently, but didn't find anything in, what I thought was the correct, documentation page.

Anyway, I tried attaching to the serial port using `sudo screen /dev/ttyACM2`, but seems to break something. The light on the ST-link blinks so fast, it's pretty much perceived as "always on", the light on my seeeduino xiao rp2040 doesn't blink either, gdb gets stuck (e.g. interrupts are not received by the debugger) and the only output I get is the code ran before my `loop` (">>>>>>>>>>>>>>>.."). **Hovever**: I've found if I don't do `redirect_stdout`, the output from `hprintln` (which is how one prints to semihosting, using the above-mentioned rust library), is just added to the gdp output. Which first of all works as expected, and is much more convenient anyway lol. Still... am I doing something wrong?

Here is the code I'm using:


```rs
    hprintln!(">>>>>>>>>>>>>>>..").unwrap();

    let mut led_pin = pins.gpio25.into_push_pull_output();

    loop {
        led_pin.set_high().unwrap();
        hprintln!("high").unwrap();
        delay.delay_ms(500);
        led_pin.set_low().unwrap();
        hprintln!("low").unwrap();
        delay.delay_ms(500);
    }
```

Output from the second serial interface (with `redirect_stdout`):

(yes, there is something wrong with the newlines and I'm not good enough to figure it out lol)

```
                                                                                                                                                                            >>>>>>>>>>>>>>>..
         >>>>>>>>>>>>>>>..
                          >>>>>>>>>>>>>>>..
                                           >>>>>>>>>>>>>>>..
                                                            >>>>>>>>>>>>>>>..
                                                                             >>>>>>>>>>>>>>>..
                                                                                              >>>>>>>>>>>>>>>..
                                                                                                               >>>>>>>>>>>>>>>..
                                                                                                                                >>>>>>>>>>>>>>>..
                                                                                                                                                 >>>>>>>>>>>>>>>..
                                                                                                                                                                  >>>>>>>>>>>>>>>..
                                                                                                                                                                                   >>>>>>>>>>>>>>>..
                                                                                                                                                                                                    >>>>>>>>>>>>>>>..
                                                                                                                                                                                                                     >>>>>>>>>>>>>>>..
              >>>>>>>>>>>>>>>..
                               >>>>>>>>>>>>>>>..
                                                >>>>>>>>>>>>>>>..
                                                                 >>>>>>>>>>>>>>>..
                                                                                  >>>>>>>>>>>>>>>..
                                                                                                   >>>>>>>>>>>>>>>..
                                                                                                                    >>>>>>>>>>>>>>>..
                                                                                                                                     >>>>>>>>>>>>>>>..
                                                                                                                                                      >>>>>>>>>>>>>>>..
```

Output from gdb (without `redirect_stdout`):

```
(gdb) continue
Continuing.
>>>>>>>>>>>>>>>..
high
low
high
low
high
low
high
low
high
low
high
low
high
low
high
low
high
```

**Reply by dragonmux** [reply]
Unsure what's going on when you connect to the second serial port unless /dev/ttyACM2 is actually the GDB port (this is why we advise to use the project udev rules as the enumeration of the ttyACM devices is non-stable, but the udev rules provide stable names in ttyBmpGdb and ttyBmpTarg (+ versions with the serial number appended) to prevent issues.)

What you're doing looks right to us but we don't know Rust well nor the crate you're using at all. With the newlines, that's probably that the crate is only putting out one of `'\r'` and `'\n'`, but screen needs both to function properly.

**Reply by Elias-Graf** [reply]
Pretty sure the GDP port is `/dev/ttyACM1`, since I'm actively connected to that, and it's working. Yes, I've read about the udev rules, but I didn't want to introduce even more new things I need to understand. Might look into it rather sooner than later though.

**Reply by dragonmux** [reply]
Interesting. Which version of the firmware are you using, just for reference? What you said about connecting to the debug serial port hanging the firmware has us concerned but it might be a bug we've fixed, or it could be something new

**Reply by Elias-Graf** [reply]
I mean, it seems like it's constantly resetting. I don't *know* if it's hanging, just seems like it. How do I figure out the firmware version? Is this of any help?:

```
(gdb) monitor
Black Magic Probe(ST-Link/v2) v1.9.0-rc1-33-g302c8b53, Hardware Version 1
Copyright (C) 2022 Black Magic Debug Project
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
```

I pretty much compiled it a few days ago.

By the way, I'm happy to assist if you want to look into this.

**Reply by dragonmux** [reply]
Yes, indeed - the version string is in that output :)

Possible constant resetting is less than ideal.. if you could do a debug build of the firmware (`make PROBE_HOST=stlink ST_BOOTLOADER=1 ENABLE_DEBUG=1`, note you may need to disable some target support in src/Makefile to make it fit by deleting the .c lines for targets you're not currently trying to debug) and monitor the output of the debug USB serial device after running `monitor debug_bmp enable` when you connect in, it might tell us what's happening if it can get messages out in time

**Reply by Elias-Graf** [reply]
Okay, here is what I did:

1. `make clean` (no idea if that's required, but it said "Nothing to be done for 'all'", which didn't seem right)
2. `make PROBE_HOST=stlink ST_BOOTLOADER=1 ENABLE_DEBUG=1`, which gave me:
   ```
   /usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: address 0x8023d94 of blackmagic.elf section `.text' is not within region `rom'
   /usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: blackmagic.elf section `.ARM.exidx' will not fit in region `rom'
   /usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: address 0x8023d94 of blackmagic.elf section `.text' is not within region `rom'
   /usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: region `rom' overflowed by 16084 bytes
   Memory region         Used Size  Region Size  %age Used
               rom:      147156 B       128 KB    112.27%
               ram:       11908 B        20 KB     58.14%
   collect2: error: ld returned 1 exit status
   make[1]: *** [Makefile:120: blackmagic.elf] Error 1
   make: *** [Makefile:17: all] Error 2
   ```
3. I assumed that this is why I had to remove some files depending on my target, so I removed files by trial and error, until I ended up removing the following:
   ```
   - efm32.c
   - lmi.c
   - kinetis.c
   - msp432.c
   - nrf51.c
   - nxpke04.c
   - ch32f1.c
   - renesas.c
   ```
4. I flashed it with `stlink-tool ./src/blackmagic.bin`

The blackmagic firmware seems to connect, but `sudo dmesg` doesn't seem to provide me with an interface (`/dev/ttyAMCX`) anymore...:

```
[46028.704793] usb 7-2: new full-speed USB device number 48 using xhci_hcd
[46028.878505] usb 7-2: New USB device found, idVendor=0483, idProduct=3748, bcdDevice= 1.00
[46028.878512] usb 7-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[46028.878515] usb 7-2: Product: STM32 STLink
[46028.878517] usb 7-2: Manufacturer: STMicroelectronics
[46028.878519] usb 7-2: SerialNumber: \x16
[46038.756569] usb 7-2: USB disconnect, device number 48
[46039.048766] usb 7-2: new full-speed USB device number 49 using xhci_hcd
[46039.222509] usb 7-2: New USB device found, idVendor=1d50, idProduct=6018, bcdDevice= 1.00
[46039.222517] usb 7-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[46039.222519] usb 7-2: Product: Black Magic Probe (ST-Link/v2) v1.9.0-rc1-33-g302c8b53-dirty
[46039.222521] usb 7-2: Manufacturer: Black Magic Debug
[46039.222523] usb 7-2: SerialNumber: 00520016530000184E534854
[46044.568513] usb 7-2: can't set config #1, error -110
```

Reverting to the firmware compiled without debug, seems to restore the old behaviour, where I'm getting:

```
[46251.598542] usb 7-2: USB disconnect, device number 53
[46252.892241] usb 7-2: new full-speed USB device number 54 using xhci_hcd
[46253.073622] usb 7-2: New USB device found, idVendor=0483, idProduct=3748, bcdDevice= 1.00
[46253.073630] usb 7-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[46253.073632] usb 7-2: Product: STM32 STLink
[46253.073634] usb 7-2: Manufacturer: STMicroelectronics
[46253.073636] usb 7-2: SerialNumber: \x16
[46254.512802] usb 7-2: USB disconnect, device number 54
[46254.788237] usb 7-2: new full-speed USB device number 55 using xhci_hcd
[46254.969626] usb 7-2: New USB device found, idVendor=1d50, idProduct=6018, bcdDevice= 1.00
[46254.969632] usb 7-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[46254.969635] usb 7-2: Product: Black Magic Probe (ST-Link/v2) v1.9.0-rc1-33-g302c8b53-dirty
[46254.969637] usb 7-2: Manufacturer: Black Magic Debug
[46254.969638] usb 7-2: SerialNumber: 00520016530000184E534854
[46255.071681] cdc_acm 7-2:1.0: ttyACM0: USB ACM device
[46255.077672] cdc_acm 7-2:1.2: ttyACM1: USB ACM device
```

I have the following udev rule setup, in case that is breaking anything...:

```
elias@pop-os:~/Projects/blackmagic$ cat /etc/udev/rules.d/70-st-link.rules 
# OpenMoko, Inc. Black Magic Debug Probe
ATTRS{idVendor}=="1d50", ATTRS{idProduct}=="6018", TAG+="uaccess"
```

**Reply by Elias-Graf** [reply]
The probe also hangs when trying to reconnect to a device that was reset:

```
(gdb) load
Error erasing flash with vFlashErase packet
(gdb) monitor swdp_scan
Target voltage: 3.35V
You are now detached from the previous target.
Available Targets:
No. Att Driver
 1      Raspberry RP2040 M0+
 2      Raspberry RP2040 M0+
 3      Raspberry RP2040 Rescue (Attach to reset!) 
(gdb) attach 1
A program is being debugged already.  Kill it? (y or n) y
Attaching to program: /home/elias/Projects/keyboard/target/thumbv6m-none-eabi/debug/keyboard, Remote target

Program stopped.
Ignoring packet error, continuing...
Ignoring packet error, continuing...
```

Again, the LED on the ST-Link is blue (instead of red or blinking when doing something), and the GDP prompt is unresponsive.

**Reply by dragonmux** [reply]
> Okay, here is what I did:
> 
>     1. `make clean` (no idea if that's required, but it said "Nothing to be done for 'all'", which didn't seem right)

If you were working from a fresh clone or having done `git clean`, then that makes sense - from the output below it looks fine.

> 
>     2. `make PROBE_HOST=stlink ST_BOOTLOADER=1 ENABLE_DEBUG=1`, which gave me:
>        ```
>        /usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: address 0x8023d94 of blackmagic.elf section `.text' is not within region `rom'
>        /usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: blackmagic.elf section `.ARM.exidx' will not fit in region `rom'
>        /usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: address 0x8023d94 of blackmagic.elf section `.text' is not within region `rom'
>        /usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: region `rom' overflowed by 16084 bytes
>        Memory region         Used Size  Region Size  %age Used
>                    rom:      147156 B       128 KB    112.27%
>                    ram:       11908 B        20 KB     58.14%
>        collect2: error: ld returned 1 exit status
>        make[1]: *** [Makefile:120: blackmagic.elf] Error 1
>        make: *** [Makefile:17: all] Error 2
>        ```
> 
>     3. I assumed that this is why I had to remove some files depending on my target, so I removed files by trial and error, until I ended up removing the following:
>        ```
>        - efm32.c
>        - lmi.c
>        - kinetis.c
>        - msp432.c
>        - nrf51.c
>        - nxpke04.c
>        - ch32f1.c
>        - renesas.c
>        ```

That's as expected, you did the right thing here.

> 
>     4. I flashed it with `stlink-tool ./src/blackmagic.bin`
> 
> 
> The blackmagic firmware seems to connect, but `sudo dmesg` doesn't seem to provide me with an interface (`/dev/ttyAMCX`) anymore...:
> 
> ```
> [46028.704793] usb 7-2: new full-speed USB device number 48 using xhci_hcd
> [46028.878505] usb 7-2: New USB device found, idVendor=0483, idProduct=3748, bcdDevice= 1.00
> [46028.878512] usb 7-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
> [46028.878515] usb 7-2: Product: STM32 STLink
> [46028.878517] usb 7-2: Manufacturer: STMicroelectronics
> [46028.878519] usb 7-2: SerialNumber: \x16
> [46038.756569] usb 7-2: USB disconnect, device number 48
> [46039.048766] usb 7-2: new full-speed USB device number 49 using xhci_hcd
> [46039.222509] usb 7-2: New USB device found, idVendor=1d50, idProduct=6018, bcdDevice= 1.00
> [46039.222517] usb 7-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
> [46039.222519] usb 7-2: Product: Black Magic Probe (ST-Link/v2) v1.9.0-rc1-33-g302c8b53-dirty
> [46039.222521] usb 7-2: Manufacturer: Black Magic Debug
> [46039.222523] usb 7-2: SerialNumber: 00520016530000184E534854
> [46044.568513] usb 7-2: can't set config #1, error -110
> ```

This isn't so good though - this indicates something having gone dreadfully wrong. The USB configuration has to be set before any other part of enumeration (like the CDC ACM interfaces being bound by the driver) can occur at that point. We're interested on what could have gone on here as it's not the first time people have reported issues with ST-Link v2's trying to run debug builds. If you have, for example, a USB analyser this would be a useful thing to get a capture from.

> 
> Reverting to the firmware compiled without debug, seems to restore the old behaviour, where I'm getting:
> 
> ```
> [46251.598542] usb 7-2: USB disconnect, device number 53
> [46252.892241] usb 7-2: new full-speed USB device number 54 using xhci_hcd
> [46253.073622] usb 7-2: New USB device found, idVendor=0483, idProduct=3748, bcdDevice= 1.00
> [46253.073630] usb 7-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
> [46253.073632] usb 7-2: Product: STM32 STLink
> [46253.073634] usb 7-2: Manufacturer: STMicroelectronics
> [46253.073636] usb 7-2: SerialNumber: \x16
> [46254.512802] usb 7-2: USB disconnect, device number 54
> [46254.788237] usb 7-2: new full-speed USB device number 55 using xhci_hcd
> [46254.969626] usb 7-2: New USB device found, idVendor=1d50, idProduct=6018, bcdDevice= 1.00
> [46254.969632] usb 7-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
> [46254.969635] usb 7-2: Product: Black Magic Probe (ST-Link/v2) v1.9.0-rc1-33-g302c8b53-dirty
> [46254.969637] usb 7-2: Manufacturer: Black Magic Debug
> [46254.969638] usb 7-2: SerialNumber: 00520016530000184E534854
> [46255.071681] cdc_acm 7-2:1.0: ttyACM0: USB ACM device
> [46255.077672] cdc_acm 7-2:1.2: ttyACM1: USB ACM device
> ```
> 
> I have the following udev rule setup, in case that is breaking anything...:
> 
> ```
> elias@pop-os:~/Projects/blackmagic$ cat /etc/udev/rules.d/70-st-link.rules 
> # OpenMoko, Inc. Black Magic Debug Probe
> ATTRS{idVendor}=="1d50", ATTRS{idProduct}=="6018", TAG+="uaccess"
> ```

udev rules should not have participated in this. The failure is occurring far too early for them to be involved. We do suggest using the project rules files, but there's nothing particularly wrong with that rule.

**Reply by Elias-Graf** [reply]
Sorry, quite a lot of work during the week, so here a late response.

> This isn't so good though - this indicates something having gone dreadfully wrong. The USB configuration has to be set before any other part of enumeration (like the CDC ACM interfaces being bound by the driver) can occur at that point. We're interested on what could have gone on here as it's not the first time people have reported issues with ST-Link v2's trying to run debug builds. If you have, for example, a USB analyser this would be a useful thing to get a capture from.

Is a USB analyzer a special piece of hardware that I need to buy, or is that a software? In either case, could you give me some examples or links? If it's not too expensive, I'm willing to get one :)

**Reply by dragonmux** [reply]
Turns out that you probably got bitten by the semihosting bugs that have now been fixed in latest `main` and the latest release of the firmware.

**Comment by dragonmux** [comment]
You make a fair point regarding documenting semihosting better. We agree the documentation is lacking and it's on our list of things to fix, if someone doesn't contribute better documentation first.

At the moment that page is more focused on how to enable use of semihosting and is C-focused (the library calls talked about are specific to newlib and C family languages). It being rewritten to have sections on making use of semihosting within BMP, how to enable semihosting in C-based firmware and how to enable semihosting in Rust-based firmware at a minimum would probably be ideal and the direction we'd take it when we get around to it.
